[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format

LucasLLC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120128
* #120127
* __->__ #119816

as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

LucasLLC committed:
[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format

as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fegin commented on torch/distributed/checkpoint/format_utils.py:
```diff
@@ -54,6 +68,161 @@ def set_up_planner(
+        return Metadata(state_dict_metadata={})
+
+    def read_data(self, plan: LoadPlan, planner: LoadPlanner) -> Future[None]:
+        """
```
One optimization is that  first rank on each host to load the file and do intra-host broadcast. This will be a faster approach. We can add a todo here to revisit this feature when performance is the concern.

LucasLLC replied:
Great idea, thanks!

fegin commented on torch/distributed/checkpoint/format_utils.py:
```diff
@@ -54,6 +68,161 @@ def set_up_planner(
+                    f"At this time {type(self).__name__} only supports loading Tensors."
+                )
+
+            size = value.size()
+            state_dict_metadata[key] = TensorStorageMetadata(
+                TensorProperties(dtype=value.dtype),
+                size,
+                [ChunkStorageMetadata(offsets=torch.Size([0] * len(size)), sizes=size)],
+            )
```
Is this correct? What if the value is a DTensor?

fegin reviewed (commented):
I have some questions about DynamicMetaLoadPlanner. BroadcastingTorchSaveReader looks good to me.

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fegin reviewed (approved):
LGTM, but the CI failure is real

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] Adds storage reader and planner classes for online loading/sharding of models in torch.save format"


as title

Differential Revision: [D53718041](https://our.internmc.facebook.com/intern/diff/D53718041/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

LucasLLC commented:
@pytorchbot merge
